llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -3975,6 +3975,26 @@ InstructionCost AArch64TTIImpl::getScalarizationOverhead(
   return DemandedElts.popcount() * (Insert + Extract) * VecInstCost;
 }
 
+std::optional<InstructionCost> AArch64TTIImpl::getFP16BF16PromoteCost(
+    Type *Ty, TTI::TargetCostKind CostKind, TTI::OperandValueInfo Op1Info,
+    TTI::OperandValueInfo Op2Info, bool IncludeTrunc,
+    std::function<InstructionCost(Type *)> InstCost) const {
+  if ((ST->hasFullFP16() || !Ty->getScalarType()->isHalfTy()) &&
+      !Ty->getScalarType()->isBFloatTy())
+    return std::nullopt;
```

Comment:
this really makes my head hurt, if I understand correctly this is equivalent to:

```
if (!IsHalfTy && !IsBFloatTy)
  return std::nullopt;

if (HasFullFP16 && IsHalfTy)
  return std::nullopt;
```

?